File: app/src/main/java/com/thewizrd/simpleweather/adapters/DetailsItemGridAdapter.java
package com.thewizrd.simpleweather.adapters;

import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;

import com.google.common.collect.Iterables;
import com.thewizrd.shared_resources.controls.DetailItemViewModel;
import com.thewizrd.shared_resources.helpers.ContextUtils;
import com.thewizrd.simpleweather.controls.DetailCard;

import java.util.ArrayList;
import java.util.List;

public class DetailsItemGridAdapter extends BaseAdapter {
    private List<DetailItemViewModel> mDataset;

    @Override
    public int getCount() {
        return mDataset != null ? mDataset.size() : 0;
    }

    @Override
    public Object getItem(int position) {
        return mDataset != null ? mDataset.get(position) : null;
    }

    @Override
    public long getItemId(int position) {
        return mDataset != null && mDataset.size() > 0 ? mDataset.get(position).getDetailsType().getValue() : 0;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        DetailCard view = convertView != null ? (DetailCard) convertView : new DetailCard(parent.getContext());
        view.setStrokeWidth((int) ContextUtils.dpToPx(parent.getContext(), 1));

        view.bindModel(mDataset.get(position));

        return view;
    }

    public void updateItems(final List<DetailItemViewModel> dataset) {
        if (mDataset == null || !Iterables.elementsEqual(mDataset, dataset)) {
            mDataset = new ArrayList<>(dataset);
            notifyDataSetChanged();
        }
    }

    @Override
    public boolean hasStableIds() {
        return true;
    }
}

File: app/src/main/java/com/thewizrd/simpleweather/adapters/DetailsItemGridAdapter.kt
package com.thewizrd.simpleweather.adapters

import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import com.thewizrd.common.controls.DetailItemViewModel
import com.thewizrd.common.controls.WeatherDetailsType
import com.thewizrd.shared_resources.utils.sequenceEqual
import com.thewizrd.simpleweather.controls.DetailCard
import com.thewizrd.simpleweather.preferences.FeatureSettings

class DetailsItemGridAdapter : BaseAdapter() {
    private var mDataset: List<DetailItemViewModel>? = null

    override fun getCount(): Int {
        return mDataset?.size ?: 0
    }

    override fun getItem(position: Int): Any? {
        return mDataset?.get(position)
    }

    override fun getItemId(position: Int): Long {
        return mDataset?.get(position)?.detailsType?.value?.toLong() ?: 0L
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val view = convertView as? DetailCard ?: DetailCard(parent.context)

        view.bindModel(mDataset!![position])

        return view
    }

    fun updateItems(dataset: Collection<DetailItemViewModel>?) {
        if (mDataset == null || !sequenceEqual(mDataset, dataset)) {
            mDataset = dataset?.filterNot {
                ((it.detailsType == WeatherDetailsType.SUNRISE || it.detailsType == WeatherDetailsType.SUNSET) && FeatureSettings.isSunPhaseEnabled) ||
                        (it.detailsType == WeatherDetailsType.MOONRISE || it.detailsType == WeatherDetailsType.MOONSET || it.detailsType == WeatherDetailsType.MOONPHASE) && FeatureSettings.isMoonPhaseEnabled
            }
            notifyDataSetChanged()
        }
    }

    override fun hasStableIds(): Boolean {
        return true
    }
}